src/consensus/dkg.rs

```diff
@@ -473,12 +412,78 @@ impl Participant {
     }
 }
 
-// Data for a DKG observer.
-#[derive(Default)]
-struct Observer {
-    elders_info: EldersInfo,
-    section_key_index: u64,
-    accumulator: HashMap<Result<bls::PublicKey, ()>, HashSet<XorName>>,
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
+pub(crate) struct DkgFailureProof {
+    public_key: PublicKey,
+    signature: Signature,
+}
+
+impl DkgFailureProof {
+    fn new(keypair: &Keypair, dkg_key: &DkgKey) -> Self {
+        Self {
+            public_key: keypair.public,
+            signature: crypto::sign(&failure_proof_hash(dkg_key), keypair),
+        }
+    }
+
+    fn verify(&self, dkg_key: &DkgKey) -> bool {
+        let hash = failure_proof_hash(dkg_key);
+        self.public_key.verify(&hash, &self.signature).is_ok()
+    }
+}
+
+#[derive(Default, Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
+pub(crate) struct DkgFailureProofSet(Vec<DkgFailureProof>);
+
+impl DkgFailureProofSet {
+    fn insert(&mut self, proof: DkgFailureProof) -> bool {
+        if self
+            .0
+            .iter()
+            .all(|existing_proof| existing_proof.public_key != proof.public_key)
+        {
+            self.0.push(proof);
+            true
+        } else {
+            false
+        }
+    }
+
+    fn has_agreement(&self, elders_info: &EldersInfo) -> bool {
+        self.0.len() >= majority(elders_info.elders.len())
```

Comment:
according to the below `verify` function, it seems the failures could be for different elders_info`,
so, shall we here only filter out the failure regarding this elders_info ?

Reply:
This function assumes the proofs are already valid. Will add a comment clarifying this.